File: apps/web/src/lib/seeds.ts
/**
 * Seed catalogue + clone-on-open client.
 *
 * The shelf step of /join lists books from this catalogue. Tapping one
 * clones it into the user's library (or returns their existing clone if
 * they've opened it before).
 */
import { api } from "./api";
import type { Book, BookStatus } from "./books";

export interface Seed {
  slug: string;
  title: string;
  author: string;
  source_language: string;
  topics: string[];
  attribution: string;
  /** User's personal clone if they've already opened this seed; null otherwise. */
  clone_id: string | null;
  clone_status: BookStatus | null;
}

export async function listSeeds(): Promise<Seed[]> {
  return api<Seed[]>("/seeds");
}

/** Idempotent — returns the user's existing clone if they've already opened
 *  this seed, otherwise materialises a new one (Book row + chunk copy). */
export async function cloneSeed(slug: string): Promise<Book> {
  return api<Book>(`/seeds/${encodeURIComponent(slug)}/clone`, {
    method: "POST",
  });
}

File: apps/api/app/schemas/seeds.py
"""Pydantic schemas for the seed catalogue + clone-on-open flow."""
from __future__ import annotations

import uuid

from pydantic import BaseModel

from app.models.book import BookStatus


class SeedRead(BaseModel):
    """One entry in the seed catalogue, with per-user clone state.

    Returned by ``GET /seeds``. The frontend uses ``clone_id`` to decide:

    * ``null`` → user has not opened this book yet; tapping calls
      ``POST /seeds/{slug}/clone`` to materialise a personal copy.
    * not null → user already cloned; tapping deep-links straight to
      ``/library/{clone_id}``.
    """
    slug: str
    title: str
    author: str
    source_language: str
    topics: list[str]
    attribution: str
    # Per-user clone state — populated after the visitor opens the seed.
    clone_id: uuid.UUID | None = None
    clone_status: BookStatus | None = None
